### ISSUE TITLE
error using nn.MSELoss() with basic_train.py

### ISSUE THREAD

**Issue by mc-robinson** [fastai/fastai#1863]
**Describe the bug**
`nn.MSELoss()` does not currently work as a  `learner.fit()` loss function

**Provide your installation details**
```
=== Software === 
python       : 3.6.8
fastai       : 1.0.51.dev0
fastprogress : 0.1.20
torch        : 1.0.1.post2
torch cuda   : None / is **Not available** 

=== Hardware === 
No GPUs available 
```

**To Reproduce**
GIST: Note that this mainly reproduces the relevant PyTorch issue.
https://gist.github.com/mc-robinson/3c5a068499344ec99bff6de64933586c

The main issue is here 
```python
def train_epoch(model:nn.Module, dl:DataLoader, opt:optim.Optimizer, loss_func:LossFunction)->None:
    "Simple training of `model` for 1 epoch of `dl` using optim `opt` and loss function `loss_func`."
    model.train()
    for xb,yb in dl:
        loss = loss_func(model(xb), yb)
        loss.backward()
        opt.step()
        opt.zero_grad()
```
in https://github.com/fastai/fastai/blob/5ac8dfe800dddb6ede6ce4178a1d68de7b087f6e/fastai/basic_train.py . `model(xb)` has shape (N,1) while `yb` has shape (N,). In this case, the DataLoader is returning the shapes that cause this function not to work with nn.MSELoss() due to broadcasting.  See more at this pytorch issue https://github.com/pytorch/pytorch/issues/16045 .

**Additional context**
I believe the fix would be to do something like the code below for now, but I am not sure how it will effect other loss functions. One might need to check if `loss_func == nn.MSELoss()` to be sure. We will probably need to fix anything printing out a validation error:

```python
def train_epoch(model:nn.Module, dl:DataLoader, opt:optim.Optimizer, loss_func:LossFunction)->None:
    "Simple training of `model` for 1 epoch of `dl` using optim `opt` and loss function `loss_func`."
    model.train()
    for xb,yb in dl:
        if yb.ndimension() == 1:
            yb = yb.unsqueeze(1)
        loss = loss_func(model(xb), yb)
        loss.backward()
        opt.step()
        opt.zero_grad()
```

I am also trying to submit a pull request to pytorch to get the confusion fixed on that end. Hopefully that will help!

**Comment by sgugger** [comment]
Note that in fastai the default (given by data objects) is to use flattened versions of all loss functions: `MSELossFlat` for instance. In this case there's no weird broadcasting.

I don't think the basic training loops should be changed, since it might cause problems when a user ends their model with a squeeze of the output. I'm closing this here because I really think it should be fixed on the PyTorch side.
